includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -97,10 +97,13 @@ public function __construct(
 		$this->supports           = [
 			'products',
 			'refunds',
-			'tokenization',
-			'add_payment_method',
 		];
 
+		// If the setting to enable saved cards is enabled, then we should support tokenization and adding payment methods.
+		if ( $this->is_saved_cards_enabled() ) {
+			$this->supports = array_merge( $this->supports, [ 'tokenization', 'add_payment_method' ] );
+		}
+
```

Comment:
We should only call `$this->is_saved_cards_enabled()` after `$this->form_fields` has been set. Otherwise, the default value for `saved_cards` is not used and people who have updated the plugin without saving their settings would not be able to use Saved Cards, even if it's technically enabled by default. This code should be moved to after `$this->form_fields = ...` or even `$this->init_settings()`.

For more info, here's how WooCommerce Core [uses the default value](https://github.com/woocommerce/woocommerce/blob/ab4a46e32e338cbf33094ab26144744a021a21a9/includes/abstracts/abstract-wc-settings-api.php#L280-L296) in case the setting is not yet saved to the database. By calling `is_saved_cards_enabled` before `$form_fields` is set, the `get_option( 'saved_cards' )` function would return an empty string as the default value.

It looks like this is the reason for the E2E tests to be failing as well.

Reply:
Thanks Luiz, great spot!